test/unittests/CandidateTest.php

```diff
@@ -593,6 +593,74 @@ public function testGetValidSubprojectsReturnsEmptyArray(): void
         $this->assertEquals($this->_candidate->getValidSubprojects(), array());
     }
 
+    /**
+     * Test getSubprojectForMostRecentVisit returns most recent visit's label
+     *
+     * @covers Candidate::getSubprojectForMostRecentVisit
+     * @return void
+     */
+    public function testGetSubprojectForMostRecentVisitReturnsMostRecentVisitLabel()
+    {
+        $subproject = array(
+            array(
+                'SubprojectID' => 1,
+                'title'        => 'testSubproject' 
+            )
+        );
+
+        $this->_setUpTestDoublesForSelectCandidate();
+
+        $this->_dbMock->expects($this->any())
+            ->method('pselect')
+            ->with(
+                $this->stringContains(
+                   "SELECT SubprojectID, title 
+                    FROM subproject 
+                    WHERE SubprojectID = :subprj"
+                )
+            )
+            ->willReturn(
+                $subproject
+            );
+
+        $expectedSubproject = array(
+            'SubprojectID' => 1,
+            'title'        => 'testSubproject'
+        );
+
+        $this->_candidate->select($this->_candidateInfo['CandID']);
+        $this->assertEquals(
+            $expectedSubproject,
+            $this->_candidate->getSubprojectForMostRecentVisit()
+        );
+    }
+
+    /**
+     * Test getSubprojectForMostRecentVisit returns null if there is no visit with a Date_visit
+     *
+     * @covers Candidate::getSubprojectForMostRecentVisit
+     * @return void
+     */
+    public function testGetSubprojectForMostRecentVisitReturnsNull()
+    {
+        $subproject = array();
+        $this->_setUpTestDoublesForSelectCandidate();
+        
+        $this->_dbMock->expects($this->any())
+            ->method('pselect')
+            ->with(
+                $this->stringContains(
+                   "SELECT SubprojectID, title 
+                    FROM subproject 
+                    WHERE SubprojectID = :subprj"
+                )
+            )
+            ->willReturn($subproject);
+
+        $this->_candidate->select($this->_candidateInfo['CandID']);
```

Comment:
Here , need a CandID type. not a array.

Reply:
hi @kongtiaowang ,  could you add direct 'suggested changes' to the code? i'm not quite sure what I need to change here

Reply:
hi @kongtiaowang can you explain what you mean by CandID type, not array?